lib/rpmsg/rpmsg_core.c

```diff
@@ -174,17 +174,28 @@ void *rpmsg_get_tx_buffer(struct rpmsg_virtio_device *rvdev,
 {
 	void *data;
 
+#if !defined VIRTIO_SLAVE_ONLY  && !defined  VIRTIO_MASTER_ONLY
 	if (rpmsg_virtio_get_role(rvdev) == RPMSG_MASTER) {
 		data = virtqueue_get_buffer(rvdev->svq, (uint32_t *)len, idx);
 		if (data == RPMSG_NULL) {
 			data = sh_mem_get_buffer(rvdev->shbuf);
 			*len = RPMSG_BUFFER_SIZE;
 		}
 	} else {
-		data =
-		    virtqueue_get_available_buffer(rvdev->svq, idx,
-						   (uint32_t *)len);
+		data = virtqueue_get_available_buffer(rvdev->svq, idx,
+						      (uint32_t *)len);
 	}
+#else
+#if defined VIRTIO_MASTER_ONLY
+	data = virtqueue_get_buffer(rvdev->svq, (uint32_t *)len, idx);
+	if (data == RPMSG_NULL) {
+		data = sh_mem_get_buffer(rvdev->shbuf);
+		*len = RPMSG_BUFFER_SIZE;
+	}
+#else
+	data = virtqueue_get_available_buffer(rvdev->svq, idx, (uint32_t *)len);
+#endif
+#endif
```

Comment:
we have three implementations here, i think we should just have one implementation, and allows part of it disabled if the feature is disabled